File: client/py/rana_socket/commands.py
"""Type-safe FlatBuffers command builders + framed socket transport.

One builder method per command `key` — the client cannot lie about the command
key it sends. Every command maps onto the generic envelope Command{key, action,
target, params, data} (see schema/command.fbs); the daemon validates `key`
against [commands.allowed] and routes to scripts/<key>.pluto.

The generated FlatBuffers module (`command_generated.py`) ships inside this
package and is produced by:

    flatc --gen-onefile --python -o schema/generated schema/command.fbs
"""

import base64
import logging
import socket
import struct
import time

import flatbuffers

from . import command_generated as fb

logger = logging.getLogger(__name__)


STATUS_NAMES = {
    fb.Status.OK: "OK",
    fb.Status.UNKNOWN_COMMAND: "UNKNOWN_COMMAND",
    fb.Status.EXECUTION_FAILED: "EXECUTION_FAILED",
    fb.Status.INVALID_PAYLOAD: "INVALID_PAYLOAD",
    fb.Status.FORBIDDEN: "FORBIDDEN",
    fb.Status.SAFETY_GATE_FAILED: "SAFETY_GATE_FAILED",
    fb.Status.FORWARD_TO_CLIENT: "FORWARD_TO_CLIENT",
}


def _str(builder, s):
    return builder.CreateString(s) if s else 0


def _params(builder, params):
    """Build a [StrPair] vector from a list of (k, v) string tuples.

    Uses the raw StartVector/PrependUOffsetTRelative/EndVector primitives
    instead of the generated CommandCreateParamsVector (which maps to
    Builder.CreateVectorOfTables — a method that exists only in unreleased
    flatbuffers master, not in any released Python runtime).
    """
    if not params:
        return 0
    offs = []
    for k, v in params:
        koff = builder.CreateString(k)
        voff = builder.CreateString(v)
        StrPairStart = fb.StrPairStart
        StrPairStart(builder)
        fb.StrPairAddK(builder, koff)
        fb.StrPairAddV(builder, voff)
        offs.append(fb.StrPairEnd(builder))
    fb.CommandStartParamsVector(builder, len(offs))
    for o in reversed(offs):
        builder.PrependUOffsetTRelative(o)
    return builder.EndVector(len(offs))


def _envelope(builder, key, action="", target="", params=None, data=None,
              request_id: int = 0) -> bytes:
    """Finish a generic Command envelope."""
    key_off = builder.CreateString(key)
    action_off = _str(builder, action)
    target_off = _str(builder, target)
    params_off = _params(builder, params or [])
    data_off = builder.CreateByteVector(bytes(data)) if data else 0

    fb.CommandStart(builder)
    fb.CommandAddRequestId(builder, request_id)
    fb.CommandAddKey(builder, key_off)
    fb.CommandAddAction(builder, action_off)
    fb.CommandAddTarget(builder, target_off)
    if params_off:
        fb.CommandAddParams(builder, params_off)
    if data_off:
        fb.CommandAddData(builder, data_off)
    cmd = fb.CommandEnd(builder)
    builder.Finish(cmd)
    return bytes(builder.Output())


class CommandBuilder:
    """Type-safe builders — one method per command key."""

    @staticmethod
    def speak(text: str, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(256)
        return _envelope(b, "speak", action="speak", target=text, request_id=request_id)

    @staticmethod
    def ask(text: str, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(256)
        return _envelope(b, "ask", target=text, request_id=request_id)

    @staticmethod
    def light(action: str, room: str = "all", request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(64)
        return _envelope(b, "lights", action=action, target=room, request_id=request_id)

    @staticmethod
    def volume(level: int, mute: bool = False, request_id: int = 0) -> bytes:
        assert 0 <= level <= 100
        b = flatbuffers.Builder(64)
        params = [("mute", "true")] if mute else []
        return _envelope(b, "volume", action="set", target=str(level),
                         params=params, request_id=request_id)

    @staticmethod
    def power(action: str, confirm: bool, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(64)
        params = [("confirm", "true")] if confirm else []
        return _envelope(b, "power", action=action, params=params, request_id=request_id)

    @staticmethod
    def launch_browser(url=None, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(64)
        return _envelope(b, "browser", action="open", target=url or "", request_id=request_id)

    @staticmethod
    def play_movie(path: str, start_seconds: int = 0, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(256)
        params = [("start", str(start_seconds))] if start_seconds else []
        return _envelope(b, "play_movie", action="play", target=path,
                         params=params, request_id=request_id)

    @staticmethod
    def media(action: str, target=None, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(64)
        return _envelope(b, "media", action=action, target=target or "", request_id=request_id)

    @staticmethod
    def lookup_machine(subnet=None, timeout_ms: int = 0, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(64)
        params = [("timeout_ms", str(timeout_ms))] if timeout_ms else []
        return _envelope(b, "lookup_machine", action="lookup", target=subnet or "",
                         params=params, request_id=request_id)

    @staticmethod
    def mc_server(action: str, world_name=None, request_id: int = 0) -> bytes:
        b = flatbuffers.Builder(64)
        return _envelope(b, "mc_server", action=action, target=world_name or "",
                         request_id=request_id)

    @staticmethod
    def talk(audio: bytes, encoding: str = "pcm16", sample_rate: int = 16000,
             channels: int = 1, request_id: int = 0) -> bytes:
        """Build a `talk` Command: raw audio bytes + format metadata. The daemon
        decodes/transcribes on-server (rana-serializer + STT hop), then routes the
        transcript through the ask hop — exactly as `ask()` does with text. The
        (untrusted) client never sees the LLM or STT."""
        b = flatbuffers.Builder(1024)
        params = [
            ("encoding", encoding),
            ("sample_rate", str(sample_rate)),
            ("channels", str(channels)),
        ]
        return _envelope(b, "talk", action=encoding, target="", params=params,
                         data=bytes(audio), request_id=request_id)


def _recv_exact(sock: socket.socket, n: int) -> bytes:
    buf = b""
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("connection closed mid-frame")
        buf += chunk
    return buf


def send(sock: socket.socket, payload: bytes) -> dict:
    """Sends one framed command, returns the daemon's decoded response.

    A response may instead be a forward directive: status FORWARD_TO_CLIENT
    (7) means the daemon could not run the command locally and has base64-
    encoded the inner Command in the message for the caller to relay elsewhere.
    """
    logger.debug("send: transmitting %d-byte framed command", len(payload))
    t0 = time.monotonic()
    try:
        sock.sendall(struct.pack(">I", len(payload)) + payload)
        length = struct.unpack(">I", _recv_exact(sock, 4))[0]
        resp_bytes = _recv_exact(sock, length)
    except (OSError, ConnectionError) as e:
        logger.error("send: transport error after %.2fs: %s", time.monotonic() - t0, e)
        raise
    r = fb.Response.GetRootAsResponse(resp_bytes, 0)
    status = r.Status()
    dt = time.monotonic() - t0
    if status == fb.Status.FORWARD_TO_CLIENT:
        logger.info(
            "recv: %d-byte response in %.2fs — status=FORWARD_TO_CLIENT (server "
            "defers to local daemon), inner command %d bytes",
            length, dt, len(r.Message() or b""),
        )
        return {
            "request_id": r.RequestId(),
            "status": status,
            "status_name": "FORWARD_TO_CLIENT",
            "forward": True,
            "command": base64.b64decode(r.Message()),
        }
    msg = (r.Message() or b"").decode(errors="replace")
    logger.info(
        "recv: %d-byte response in %.2fs — status=%s(%d)%s",
        length, dt, STATUS_NAMES.get(status, str(status)), status,
        f" message={msg!r}" if msg else "",
    )
    return {
        "request_id": r.RequestId(),
        "status": status,
        "status_name": STATUS_NAMES.get(status, str(status)),
        "message": msg,
        "forward": False,
    }

File: client/cpp/src/commands.cpp
// FlatBuffer command construction. See commands.hpp.
//
// The daemon speaks ONE generic envelope: Command{key, action, target, params,
// data}. Each CLI verb maps onto that envelope (key == script registry name).

#include "commands.hpp"

#include <command_generated.h>

#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <string_view>
#include <vector>

namespace {

using ParamList = std::vector<std::pair<std::string, std::string>>;

// Build a generic Command envelope from its parts.
flatbuffers::Offset<rana::Command> make_command(
    flatbuffers::FlatBufferBuilder& fbb, uint64_t request_id,
    const std::string& key, const std::string& action, const std::string& target,
    const ParamList& params, const std::string& data) {
    const auto k = fbb.CreateString(key);
    const auto a = fbb.CreateString(action);
    const auto t = fbb.CreateString(target);

    std::vector<flatbuffers::Offset<rana::StrPair>> param_offs;
    param_offs.reserve(params.size());
    for (const auto& kv : params) {
        param_offs.push_back(
            rana::CreateStrPair(fbb, fbb.CreateString(kv.first), fbb.CreateString(kv.second)));
    }
    const auto params_off =
        param_offs.empty() ? 0 : fbb.CreateVector(param_offs);

    const auto data_off =
        data.empty() ? 0
                     : fbb.CreateVector(reinterpret_cast<const uint8_t*>(data.data()),
                                        static_cast<uint32_t>(data.size()));

    return rana::CreateCommand(fbb, request_id, k, a, t, params_off, data_off);
}

// Reports a missing-argument usage error and returns false.
bool usage_err(const char* usage) {
    std::fprintf(stderr, "%s\n", usage);
    return false;
}

}  // namespace

bool build_command(const std::string& cmd, int argc, char** argv,
                   flatbuffers::FlatBufferBuilder& fbb, uint64_t request_id,
                   std::string& /*err*/) {
    // Helper lambdas captured below per-verb.
    auto finish = [&](const std::string& key, const std::string& action,
                      const std::string& target, const ParamList& params,
                      const std::string& data = "") {
        const auto off = make_command(fbb, request_id, key, action, target, params, data);
        fbb.Finish(off);
        return true;
    };

    if (cmd == "speak") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> speak <text>");
        return finish("speak", "speak", argv[3], {});
    }

    if (cmd == "ask") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> ask <text>");
        return finish("ask", "", argv[3], {});
    }

    if (cmd == "light") {
        if (argc < 5) return usage_err("usage: rana-socket-client <addr> light <room> <on|off|toggle>");
        const std::string_view a = argv[4];
        if (a != "on" && a != "off" && a != "toggle")
            return usage_err("light: action must be on|off|toggle");
        return finish("lights", std::string(a), argv[3], {});
    }

    if (cmd == "volume") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> volume <0-100> [mute]");
        char* end = nullptr;
        const long lvl = std::strtol(argv[3], &end, 10);
        if (end == argv[3] || *end != '\0' || lvl < 0 || lvl > 100)
            return usage_err("volume: level must be 0-100");
        const bool mute = argc >= 5 && std::string_view(argv[4]) == "mute";
        ParamList params;
        if (mute) params.emplace_back("mute", "true");
        return finish("volume", "set", std::to_string(lvl), params);
    }

    if (cmd == "power") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> power <shutdown|standby|reboot> [--confirm]");
        const std::string_view a = argv[3];
        if (a != "shutdown" && a != "standby" && a != "reboot")
            return usage_err("power: action must be shutdown|standby|reboot");
        const bool confirm = argc >= 5 && std::string_view(argv[4]) == "--confirm";
        ParamList params;
        if (confirm) params.emplace_back("confirm", "true");
        return finish("power", std::string(a), "", params);
    }

    if (cmd == "browser") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> browser <url>");
        return finish("browser", "open", argv[3], {});
    }

    if (cmd == "movie") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> movie <path> [start_seconds]");
        uint64_t start = 0;
        if (argc >= 5) {
            char* end = nullptr;
            start = std::strtoull(argv[4], &end, 10);
            if (end == argv[4]) start = 0;
        }
        ParamList params;
        if (start) params.emplace_back("start", std::to_string(start));
        return finish("play_movie", "play", argv[3], params);
    }

    if (cmd == "media") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> media <play|pause|stop|playpause> [target]");
        const std::string_view a = argv[3];
        if (a != "play" && a != "pause" && a != "playpause" && a != "stop")
            return usage_err("media: action must be play|pause|stop|playpause");
        const std::string target = argc >= 5 ? argv[4] : "";
        return finish("media", std::string(a), target, {});
    }

    if (cmd == "machine") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> machine <subnet> [timeout_ms]");
        uint32_t timeout = 3000;
        if (argc >= 5) {
            char* end = nullptr;
            const long t = std::strtol(argv[4], &end, 10);
            if (end != argv[4] && *end == '\0' && t >= 0)
                timeout = static_cast<uint32_t>(t);
        }
        ParamList params;
        params.emplace_back("timeout_ms", std::to_string(timeout));
        return finish("lookup_machine", "lookup", argv[3], params);
    }

    if (cmd == "mc-server") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> mc-server <start|stop|restart> [world]");
        const std::string_view a = argv[3];
        if (a != "start" && a != "stop" && a != "restart")
            return usage_err("mc-server: action must be start|stop|restart");
        const std::string world = argc >= 5 ? argv[4] : "";
        return finish("mc_server", std::string(a), world, {});
    }

    if (cmd == "talk") {
        if (argc < 4) return usage_err("usage: rana-socket-client <addr> talk <audio-file> [encoding=wav|pcm16] [sample_rate] [channels]");
        std::FILE* fp = std::fopen(argv[3], "rb");
        if (!fp) { std::perror("talk: open"); return false; }
        std::fseek(fp, 0, SEEK_END);
        const long size = std::ftell(fp);
        std::fseek(fp, 0, SEEK_SET);
        if (size <= 0) { std::fclose(fp); return usage_err("talk: empty audio file"); }
        std::string raw(static_cast<size_t>(size), '\0');
        if (std::fread(&raw[0], 1, raw.size(), fp) != raw.size()) {
            std::fclose(fp); std::perror("talk: read"); return false;
        }
        std::fclose(fp);

        const std::string_view enc = argc >= 5 ? argv[4] : "wav";
        uint32_t sample_rate = 0;
        uint8_t channels = 1;
        if (argc >= 6) {
            char* end = nullptr;
            const long sr = std::strtol(argv[5], &end, 10);
            if (end != argv[5] && *end == '\0' && sr >= 0) sample_rate = static_cast<uint32_t>(sr);
        }
        if (argc >= 7) {
            char* end = nullptr;
            const long ch = std::strtol(argv[6], &end, 10);
            if (end != argv[6] && *end == '\0' && ch > 0 && ch <= 255)
                channels = static_cast<uint8_t>(ch);
        }
        ParamList params;
        params.emplace_back("encoding", std::string(enc));
        params.emplace_back("sample_rate", std::to_string(sample_rate));
        params.emplace_back("channels", std::to_string(channels));
        return finish("talk", std::string(enc), "", params, raw);
    }

    std::fprintf(stderr, "rana-voice-agent: unknown command '%s'\n", cmd.c_str());
    return false;
}
